### ISSUE TITLE
Support complete numpy functions and some SciPy functions

### ISSUE THREAD

**Issue by linjing-lab** [google/jax#10987]
Please:

- [ ] Check for duplicate requests.
- [ ] Describe your goal, and if possible provide a code snippet with a motivating example.
Support complete `numpy` functions and some `SciPy` functions

**Comment by jakevdp** [comment]
Thanks for the request! The unimplemented NumPy functions are tracked in https://github.com/google/jax/issues/70, along with some notes about which ones we don't intend to implement.

As for supporting some SciPy functions, I don't think we have any issue specifically tracking that. We've not created any detailed roadmap for SciPy function support, but the general tenor of the team's goal has been "support SciPy functions that are reasonable to support". Do you have any particular examples in mind of what you'd like to see supported?

**Comment by jakevdp** [comment]
Since all IO processing will happen on the host, I don't think there would be any advantage to having JAX-specific wrappers for these functions. Even if we implemented them, they would probably look something like this:
```python
def loadmat(*args):
  return jnp.array(scipy.io.loadmat(*args))
```
At that point you may as well just call scipy and convert the result to JAX manually.

With that in mind, I don't really see any benefit to adding a `jax.scipy.io` submodule, as it would add complexity with no real benefit.  What do you think?

**Comment by linjing-lab** [comment]
I think the method of supporting matlab style data in `scipy.io` is very good, and the direction and prospect of JAX framework are very good.

```python
# matfile read and write
from .matlab import loadmat, savemat, whosmat
```
I just found the corresponding method in the source code of SciPy, and the methods supported by SciPy are quite perfect. I am still exploring the methods supported by JAX. Please forgive me if I have any incorrect understanding.

> I am trying to develop a new machine learning algorithms library called `jax-ml`, which will be built and designed based on JAX framework, so I will keep communicating with you for a long time.

**Comment by jakevdp** [comment]
Sorry, I may have been unclear. My recommendation is that when you want to load a matrix as a JAX array, you use:
```python
mat = jnp.array(scipy.io.loadmat(filename))
```
There's not really any way to do better than this.

**Comment by linjing-lab** [comment]
I think this is a good way to deal with data sets. I hope more topics will be discussed and solved in the future.

**Comment by MarioAuditore** [comment]
Hello! Are there any plans on adding such functions as `logm` and `fractional_matrix_power` from `scipy.linalg`? Kinda weird that jax already has `scipy.linalg.expm` but not the rest of the functions.

**Comment by linjing-lab** [comment]
I recommend the following snippets to unify `numpy.ndarray` to `jax.numpy.array`:
```python
def logm(A, disp: bool=True):
  return jnp.array(scipy.linalg.logm(A, disp))
```
and
```python
def fractional_matrix_power(A, t: float):
  return jnp.array(scipy.linalg.fractional_matrix_power(A, t))
```
The tracked issue of unimplemented Numpy functions #70 was already closed, and I didn't find issue specifically for tracking unimplemented SciPy functions.

**Comment by MarioAuditore** [comment]
Thak you for your answer! Code snippets work fine, but they give error when I attempt to compute jacobian (due to convertion from numpy to jax)

**Comment by linjing-lab** [comment]
You may refer to the following case to construct funs computed with `jacobian`:
```python
def f(x):
   return jnp.asarray([x[0], 5*x[2], 4*x[1]**2 - 2*x[2], x[2] * jnp.sin(x[0])])
```
and compute like:
```python
import jax
from jax import numpy as jnp
print(jax.jacobian(f)(jnp.array([1., 2., 3.]))) # you can replace `jnp.array([1., 2., 3.])` with any array of `jnp.array` format.
```

**Comment by jakevdp** [comment]
> plans on adding such functions as `logm` and `fractional_matrix_power` from `scipy.linalg`? Kinda weird that jax already has `scipy.linalg.expm` but not the rest of the functions.

Hi - these functions are generally implemented on an as-needed basis (subject to having appropriate routines to compute them on at least CPU & CUDA backends). If you're interested in requesting them (or contributing an implementation!) the best way to start would be to open a specific issue for the function.

This issue is a little too broad to be useful at this point, so I think I'll close it.
